### DISCUSSION TITLE
get RocketLogger struct

### DISCUSSION THREAD

**Discussion by LuckyTurtleDev** [rwf2/Rocket#2323] [Questions]
Is it possible to get the [RocketLogger](https://github.com/SergioBenitez/Rocket/blob/7bbe0457a5ecc9d4b3ca0a2bc6120ffbe1e7febd/core/lib/src/log.rs#L64) struct manual?
I would like to use `sentry-log` to use sending errors to sentry. But using sentry as logger would total remove the local logging.

So I had though about writing a own logging crate with does pass the logs to sentry and the RocketLogger. But but in contrast to other logger I can not find I way to get the RocketLogger struct manual.

**Comment by edgarogh** [comment]
While the structure isn't public for forward-compatibility reasons, the `log`crate can give you a static reference to the current logger with [`log::logger()`](https://docs.rs/log/latest/log/fn.logger.html), as a trait object.

You build your sentry logger with [`SentryLogger::with_dest`](https://docs.rs/sentry-log/latest/sentry_log/struct.SentryLogger.html#method.with_dest) and give it the refence returned by the above function, effectively "stealing" and wrapping rocket's logger.

I haven't tested anything I'm telling you to do so I might be wrong. I still hope I answered your question.

**Reply by LuckyTurtleDev** [reply]
Hi, I do not see how this can work. Even if I can get the RocketLogger with [log::logger()](https://docs.rs/log/latest/log/fn.logger.html) the logs will still be sent to the RocketLogger and I have no option to get the logs from it or  to replace the RocketLogger with my own Logger implementation. Because the logger can not be changed anymore, if it was set once. See [log::set_logger](https://docs.rs/log/latest/log/fn.set_logger.html)

**Reply by edgarogh** [reply]
You're right, I didn't know that you couldn't replace the logger at runtime. In this case, I'm afraid you'll have to extract the `RocketLogger` completely (maybe as a separate local crate) since it is private. There really isn't any other way to get access to the structure.

**Reply by LuckyTurtleDev** [reply]
Would a pr, witch does add an option to add a second logger will be welcome?

see https://github.com/SergioBenitez/Rocket/issues/2324

**Reply by edgarogh** [reply]
Maybe, I'm not a maintainer of this project so I can't say anything.

I'm pretty sure that the `RocketLogger` will remain private, but it might be possible to ask for a custom logger factory like `impl FnOnce(&'static dyn Log) -> &'static dyn Log`. But this would not be a priority at all.

You're better off copying/reimplementing the `RocketLogger` yourself if the only thing you really want is its look.
